### ISSUE TITLE
Support for .NET core

### ISSUE THREAD

**Issue by kkozmic** [castleproject/Core#90]
What is says on the tin.

This was triggered by https://github.com/Moq/moq4/issues/168 but obviously there will me more demand for this than just from Moq.

So let's talk about it.

**Comment by smudge202** [comment]
Happy to help on this. Looks like an awful lot of work (at a glance). If we can come up with a game plan, can try to rope in more people to give us a hand...

//CC @mattridgway, @herecydev, @sblackler

**Comment by kkozmic** [comment]
thanks @smudge202 for your kind offer.

Perhaps let's start by getting the lay of the land. What makes you think it's an awful lot of work? Do you have a report from the Portability Analyser you can post?

I don't have a working Windows VM at the moment which makes it somewhat tricky for me to do this.

**Comment by jonorossi** [comment]
The report output was linked in the moq issue:
http://dotnet.github.io/port-to-core/Moq4_ApiPortabilityAnalysis.htm

**Comment by kkozmic** [comment]
Oh right. I see. At a glance it looks like a fair bit of those are actually coming from `DictionaryAdapter`. We had merged it into Castle.Core for simplicity in pre-Nuget era, but I wouldn't be too fussed to split it out if it helps us get some wins here.

**Comment by kkozmic** [comment]
any chance we can get that report grouped by usage, so we can see where the problematic APIs are being used across our codebase?

**Comment by jonorossi** [comment]
And much of the rest is CAS/security transparency, serialization/remoting, smtpclient facade, and the simplified reflection API.

For things like `GetTypeInfo()` (which is only in .NET 4.5/4.6 only) I think we'd have to implement our own extension method which would mirror the .NET 4.5 one to reduce conditional compilation. We'll have to conditionally compile DP strong name support out, as well as other things.

**Comment by kkozmic** [comment]
given we're most likely going to make this a v4.0 release, we have some room for breaking changes I suppose.

Dropping Silverlight support is something I wanted to do for a long while.

What's the issue with strong naming @jonorossi ?

**Comment by jonorossi** [comment]
.NET Core doesn't support strong naming, so the Reflection.Emit API doesn't support it, no a biggie.

**Comment by kkozmic** [comment]
gotcha

**Comment by smudge202** [comment]
So glad that didn't turn into a strong naming debate... :)

Not sure if the report can be changed to group by usage like kkozmic asked. Personally, I'd create a new vNext library targetting only dnxcore50, dump all existing cs files into the new project's directory, and start working through the error list.

In my opinion, best approach is to get the code to run against dnxcore50, then bring in net451/net40, make any changes required using ifdefs, then bring in other profiles you want to support, rinse lather repeat.

Take the TypeInfo and dnxcore code as a baseline, from which all other profiles are ifdef'd in.

**Comment by MattGal** [comment]
@smudge202 , it's actually not too bad to include existing C# files using an ASP.NET VNext project in a sub-directory of the solution (as it will pick up files adjacent / below it automatically...) and use the "code" tag to include them, setting the define constant for Silverlight (which works in many case) and defining an additional one such as CORECLR or DNXCORE50 for where the implementation has to differ from SL5.  

Duplicating all the existing sources is not fun for bug fixes / updating features.

**Comment by smudge202** [comment]
Yeh, I took the same approach with the FluentAssertions port I've been working on. Just use `compile` in `project.json` to pull in code.

**Comment by MattGal** [comment]
I've going to start working on this this week, so if anyone already has please let me know so I don't duplicate effort.  If all goes well, I should be done in a week or two.

**Comment by uhaciogullari** [comment]
Do you have any plans about migrating the unit tests to xUnit.net? Only it has a working test runner in 
dnx (or CoreCLR or whatever they are calling it now).

I think Castle Project is one best things about .NET. Let me know if I can do anything to help it work.

**Comment by uhaciogullari** [comment]
I have attemped to compile the project on DNX. I have only added a project file and referenced some NuGet packages. Castle.Core classes remain unmodified.

https://github.com/uhaciogullari/Core/tree/core

- I had VS 2015 RC and VS Code on my machine. 
- Installed DNX (or DNVM) like documented [here](https://github.com/aspnet/Home#getting-started-on-windows). Don't forget to run "dnvm upgrade" or you don't get dnu.
- Created an [ASP.NET class library with yo](https://code.visualstudio.com/Docs/ASPnet5).
- Copied the generated file to Castle.Core folder
- Executed dnu restore in that folder
- Executed dnu build and got 3290 compile errors :cold_sweat:
- At this point I started referencing system libraries but they are split into many NuGet packages.
- VS Code autocomplete doesn't help much. It probably shows the packages for full CLR.
- Opened up [NuGet Package Explorer] and start digging the feed at https://www.myget.org/F/aspnetvnext/api/v2
- Started referencing popular & sensible packages, see if they decrease the compile errors by running dnu restore & dnu build.
- Managed to get down to 1342 errors. There are a lot of Xml errors that could not be get rid of by referencing the packages. I think I tried all of them, they may have changed the namespaces. This requires code modification in cs files so I left them.
- [dnu build command was not working on Mono](https://github.com/aspnet/Home/issues/498#issuecomment-101031447) when I checked it a few weeks ago. There were a few work arounds but I didn't have the time to try it yet. Maybe some folks can try to build it from my branch and compare the output. 

Here's the last build log by the way: https://gist.github.com/uhaciogullari/ce214c172f12eb1fca0c

**Comment by jonorossi** [comment]
@uhaciogullari There are no plans to migrate from NUnit to xUnit.net. With just the Castle Core and Windsor projects we've got nearly 3000 unit tests, I don't think migrating them is a good use of time and it is likely to introduce bugs into the unit tests.

The NUnit team is aware of this: https://github.com/nunit/nunit/issues/575

**Comment by uhaciogullari** [comment]
I have managed to get down to 578 errors. Here's the last error log : https://gist.github.com/uhaciogullari/9b1762008ad36adbd071

Here's a summary of the problems I could find:

 - [Binary serialization is not supported](https://github.com/dotnet/corefx/issues/1272)
 - [Remoting is not supported](https://github.com/dotnet/corefx/issues/1677)
 - [System.Net.Mail is not supported yet](https://github.com/dotnet/corefx/issues/1006)
 - [System.Configuration](https://github.com/dotnet/corefx/issues/1348) is not supported, I guess this assembly will be replaced by Microsoft.Framework.Configuration.* packages.
 - There are some problems with Xml namespaces. I referenced every Xml package but it still can't find some XSLT types.
 - Reflection API may have changed. Many properties on Type class cause errors.

**Comment by MattGal** [comment]
This sounds about right.  I have been sidetracked for a couple weeks and not paying attention to GitHub, but I have the project fully building for Core 5.0 here: https://github.com/mattgal/Core and intend to make a tentative PR (tests are still up in the air on this... figuring out the XUnit/Nunit thing has been a challenge) tonight so folks can look at it.

**Comment by richlander** [comment]
See #92.

**Comment by MattGal** [comment]
Pinging for any comments?   It will probably be another week til I work on this some more, but I'd love to get some direction on any changes folks think are needed to take in the PR.  As it currently stands, it should not affect building the existing projects at all (if it does, that's an unintentional bug...)

**Comment by jonorossi** [comment]
@MattGal thanks for the WIP PR, however could you update it to not include a copy of all the unit tests. GItHub won't let me review it because the diff is too big:

> Showing **556 changed files** with **29,258 additions** and **193 deletions**.

**Comment by smudge202** [comment]
@jonorossi Wouldn't it be better to clone the originating branch and review locally than to remove the tests?

**Comment by jonorossi** [comment]
@smudge202 then what was the point of the pull request if I can't comment on the code inline?

**Comment by smudge202** [comment]
https://msdn.microsoft.com/en-gb/library/bb385979.aspx - article explains code annotations related to TFS, but same applies to git.

**Comment by jonorossi** [comment]
@smudge202 that is for TFVC, but yes Git has support for "blaming". I'm confused, why are you pointing out this feature?

**Comment by smudge202** [comment]
I wonder if code annotations will allow you to "comment on the code inline"?

**Comment by jonorossi** [comment]
@smudge202 the annotate/blame/praise feature in version control systems allow you to view a file's contents with the last commit displayed next to each line.

**Comment by richlander** [comment]
That's certainly unfortunate:

> @MattGal thanks for the WIP PR, however could you update it to not include a copy of all the unit tests. GItHub won't let me review it because the diff is too big

How about @MattGal rebase his PR into two, one for the code and one for the tests, but treat it as one meta PR? If they are clearly split by directory than filter-branch can do the trick.

**Comment by jonorossi** [comment]
No need for a second PR, I don't understand the hesitation of just excluding the WIP copied unit tests. You guys don't actually expect us to merge a PR that contains a copy of all the unit tests do you?

**Comment by richlander** [comment]
Help me understand. Were the unit tests changed or why are they included?

**Comment by jonorossi** [comment]
@richlander did you read @MattGal's PR description?
> A tentative set has been put in using XUnit.

If NUnit isn't going to happen for .NET Core any time soon (I see no one has engaged with Charlie, Rob and folk, https://github.com/nunit/nunit/issues/575) then moving to xUnit.net is a different discussion we'd have to entertain with that knowledge.

**Comment by MattGal** [comment]
@jonorossi I have updated the PR to not include the test projects any more, leaving them in a branch on my fork if anyone wants to use them as a basis.     I don't mind;  as is, the regular versions of Castle still build as before.   I'd like to get this in so others who use Castle more can find issues with the CoreCLR version and I can start building CoreCLR Moq4 on top of this, can you suggest what more I should do to get to that point?

**Comment by richlander** [comment]
@jonorossi I talked to @MattGal about this. @MattGal mostly clarifies it in his comment above, but it is worth capturing our conversation.

- It's a bad form to make a big shift (e.g. testing framework) w/o consultation/approval from the project maintainers.
- Currently XUnit is the only testing framework supported for CoreCLR, so the tests were/are valuable.
- We should get NUnit to support CoreCLR, too. After we finish we Moq, we can consider helping there.
- For the immediate term, the tests are valuable and should be persisted (for anyone to use), but should not be part of the PR.

**Comment by kkozmic** [comment]
Very interesting conversation guys ☺

To simplify the move how about we do it in chunks. Start off with dynamic proxy as it is the most commonly used part and leave dictionary adapter, email sender to make the move easier. Divide and conquer.

Also, risking hijacking the conversation, we should soon start looking at which frameworks versions we want to support and how many binary versions we need to produce. Perhaps a separate albeit related discussion

**Comment by richlander** [comment]
@kkozmic Are you suggesting that @MattGal should split his PR into pieces instead of taking it all (modulo the tests) at once?

I assume that your second point has more to do with how this change will affect your NuGet package. Yes?

**Comment by uhaciogullari** [comment]
Breaking changes in Reflection API leads to many changes. Many properties on Type class are now moved to TypeInfo class which can be created by calling GetTypeInfo extension method.

Should we go on with preprocessor directives or [GetTypeInfo method](https://msdn.microsoft.com/library/system.reflection.introspectionextensions.gettypeinfo%28v=VS.110%29.aspx?f=255&MSPPError=-2147217396)? It's only available on .NET 4.5 and later though. Creating the same extension method for older builds should work if you plan on supporting them. Here's a [detailed post](http://blogs.msdn.com/b/dotnet/archive/2012/08/28/evolving-the-reflection-api.aspx) about this change. It's written for WinRT but it's the same for .NET Core I think.

**Comment by jonorossi** [comment]
I've merged the `mono-support` branch (#95) into master because there was quite a lot of work there which was stable, and I wanted it in before we start getting any of this work in. We also need that better Mono support since DNX is currently using Mono on non-Windows.

Today I've also introduced two new conditional compilation symbols that will ease the migration to .NET Core. The .NET Framework builds will now define `FEATURE_SERIALIZATION` (#96) and `FEATURE_LEGACY_REFLECTION_API` (#98) ([see the README for the configuration matrix](https://github.com/castleproject/Core#conditional-compilation-symbols) which shows what builds with what). These symbols won't be defined in a .NET Core build and the intention is to continue pulling out whatever isn't supported into symbols we can enable on the full .NET Framework.

This work should help to reduce the churn of code that #92 will introduce. I don't speak for @kkozmic, but personally I think with the amount of code changed in #92 and the number of new TODOs added it is pretty hard to code review and ensure we get the changes in without causing defects.

As per the PR descriptions, the serialization symbol should be complete, however I'm only part way through the reflection API migration. I'll continue on that and continue working on getting these changes in safely.

**Comment by uhaciogullari** [comment]
@jonorossi Nice work, it fixed many compile errors on .NET Core. Without any code change it's down to 176 errors. Most of them are related to Reflection API, others are related to Remoting, Configuration and Security mainly.

Here's the build log: https://gist.github.com/uhaciogullari/d6fa16be781a905ee6d6

**Comment by kkozmic** [comment]
@richlander re: splitting the PR - yes, just floating it out there as an option if we find it's too much to do it all in one go.

re packages - it's partially about nuget but also, about effort. If nobody uses Silverlight 4 anymore, why spend time/effort keeping Castle running on sl4?

**Comment by MattGal** [comment]
Hi all,
   I now have bandwidth to basically devote full time to this for a while, but I don't mean to step on anyone's toes or duplicate effort.  It sounds like @uhaciogullari  is working on the same thing as I am.  Can I get some clarity on whether I should keep going on my efforts to make a .NET Core version of Castle?  I have a list of issues from the PR that I will start addressing although some of them have no clear answer (for instance, what to do with respect to unit tests, given NUnit is not supported on Core)

Thanks,
Matt

**Comment by davidfowl** [comment]
We're (ASP.NET team) are also pretty interested in getting Moq (and therefore Castle) up and running on CoreCLR. I just did a scan of @MattGal 's changes and they are pretty tiny, it's mostly ifdefs. Is there anything we can do to push this forward? We want to help because we actually want to use this stuff!

**Comment by uhaciogullari** [comment]
@MattGal I haven't done anything really, just created the project.json file with the correct packages and tried to compile it in DNX. I wanted to see the breaking changes in .NET API.

@davidfowl  I'm no contributer but I think you can work on your fork and publish it as an unofficial port since it's an open source project. You got your own NuGet feed for .NET Core, you also ported the tests to xUnit.  This was how it went with xUnit port as far as I remember. Contributers have to make sure that the PR meets their quality standarts and you're in a hurry.

**\<rant\>**
I can't understand why the Reflection API gets broken in every .NET subset that comes out of Microsoft. Mono and Xamarin have their own problems but at least they try to offer the same API. Did anything significant get ported to WinRT or portable libraries? The best is Silverlight and I have no idea how much work went into porting something like Castle Windsor. 

It's all the same this time. CoreCLR even has Reflection.Emit and many reflection assemblies but a basic tool like NUnit cannot be ported with ease. I really want to understand why we need the changes, I can't imagine the type system working differently in CoreCLR. You probably have very good reasons but I never saw a post about them, neither a way to deal with those changes. We are only left with broken tools.

I also wonder how developers are expected to port their applications to .NET Core. I think preprocessor directives cause a mess and porting a decent sized application will be practically impossible. Maybe we are not supposed to port them, just write new apps in .NET Core? What about our libraries? We will miss lot of them.

I don't want to sound harsh, you are the good guys who are making it better and I love your work. It's Microsoft's fault that they did nothing about cross platform issue until containers and mobile platforms forced their hand. Maybe it was worth it with Windows and SQL licences but you would expect them to be working on it since CLR boasts to be cross platform.After all these years they could have called it .NET Absolution :smile:. If the containers hit the big time (you may argue they already did) .NET will be extra uncool. 
**\</rant\>**

**Comment by davidfowl** [comment]
> @davidfowl I'm no contributer but I think you can work on your fork and publish it as an unofficial port since it's an open source project. You got your own NuGet feed for .NET Core, you also ported the tests to xUnit. This was how it went with xUnit port as far as I remember. Contributers have to make sure that the PR meets their quality standarts and you're in a hurry.

That's a pretty lame thing to do IMO. We'd rather not create unofficial forks of popular open source projects. So far we've been trying to help popular library authors port their libraries to .NET Core with varying levels of success :smile: . This issue is also "up for grabs" which means that the contributors are ok with the community taking over and helping out. We're mostly looking for guidance on how to continue, we're not trying to force a PR down your throat. If you don't like how the change was made then we'll change it, after all we run OSS projects too and understand what it means to take on changes made by community members.

**Comment by jonorossi** [comment]
Sorry for the delay in replying, I've been unwell with the flu for the last few days. I'll try to reply to your comments tomorrow.

**Comment by MattGal** [comment]
@jonorossi, some updates:
Changes:

-	Modified the .NET Core project to now be a Portable Class Library targeting .NET 4.6 + .NET Core 5.0.  This allows things like being able to see all the files in the solution (despite them being links), and building from the regular command line via MSBuild instead of needing DNX to build.  Overall the experience is improved with respect to building the library like the others are.
-	Brought back unit tests as XUnit (only way they’ll work on core) and got all but 48 passing.
-	Renamed CORECLR to NETCORE
To do: 
-	Integrate the new project into existing build scripts, document this in readme
-	Drive to 100% unit test pass rate. (currently 22 failing, 658 passing in the XUnit version of the tests.
Open questions / issues:
-	For AssemblyBundleResource / AssemblyResourceFactory, the APIs that allow for explicit load of Assemblies are gone.  There is load via AssemblyName, and when hosted in DNX there are DI interfaces for getting the same effect.  Should I cut APIs that use .Load(string), take a dependency on DNX interfaces, or other?
-	Still need guidance about what to do with respect to unit tests, as NUnit doesn’t look to be available on Core any time soon.

**Comment by MattGal** [comment]
Sorry to hear you've been ill, I am updating the PR, please give feedback when you have a moment.

**Comment by jonorossi** [comment]
Sorry I didn't get back to you on Sunday, I overestimated my recovery and the amount of other things that piled up.

@MattGal it is nice to know you've got more time to help work on this now. Your short bursts then silence for a week or more made things difficult to communicate, which was why I went off and tried making the port much easier to review than trying to get you to make those changes.

@davidfowl the changes definitely weren't small like they are now until I introduced the `FEATURE_SERIALIZATION` and `FEATURE_LEGACY_REFLECTION_API` conditional compilation symbols into master. I've hinted at it a couple of times, but my push is really to avoid another platform conditional compilation symbol and so I'm trying not to just rush a merge that'll just give us more technical debt.

**Comment by jonorossi** [comment]
@MattGal thanks for getting those bad merge changes fixed up and getting a PCL set up. The new set up looks like it should be easier to get into our build scripts.

Looking at [your branch](https://github.com/MattGal/Core/tree/Add-NetCore-Beta5/src/Castle.Core.NetCore) am I to assume that the `Castle.Core-CoreCLR` directory is no longer needed and that the PCL is in `Castle.Core.NetCore`? Regarding your question in the PR about whether it is fine for this to build with VS2015 only, yep that is fine.

Looking at your .NET Core copy of the unit tests I see they are quite out of date already. I'm working on a spike to see if we can use both testing frameworks, this will then be the precursor to us moving to xUnit.net completely or upgrading to a new NUnit with support for .NET Core in the future. I'll report back within a day or two with the results, and hopefully you'll be able to again remove your copy of the unit tests from the PR, we can't do much until we sort them out. I'd then like to get a build set up on our build server for the .NET Core target.

**Comment by MattGal** [comment]
I've squashed previous commits, deleted the CoreCLR folder, and updated the PR text.  I am committed to fixing any bugs/issues that fall out of this, but I'd like to get some more feedback (ideally with some folks trying this on VS 2015) first.

**Comment by jonorossi** [comment]
@MattGal I've made some good progress on the unit tests, looking at getting them pushed soon. I've just tried opening your branch in VS2015 after getting it installed the other day and I get the following error. I can't find where to download the PCL reference assemblies for v5.0, it [looks like it hasn't been released yet](http://stackoverflow.com/questions/31670581/migrating-portable-class-library-to-new-dnx-shared-library).

> The imported project "C:\Program Files (x86)\MSBuild\Microsoft\Portable\v5.0\Microsoft.Portable.CSharp.targets" was not found. Confirm that the path in the <Import> declaration is correct, and the fhe file exists on disk.

**Comment by jonorossi** [comment]
@MattGal I just saw a non-MSFT developer creating one of these PCL projects in another PR so I looked into my VS2015 install. Once I installed the optional "Tools and Windows SDK 10.0.10240" hidden under "Universal Windows App Development Tools" it installed the v5.0 directory and I can create and open these projects.

In your branch I can build Castle.Core.NetCore, however Castle.Core.NetCore.UnitTests doesn't build, has hundreds of errors for everything in Castle.Core not being found. Why is the new unit test project a DNX project rather than a PCL project? Did you not change this one because xunit wouldn't install? I get an "xunit.abstractions 2.0.0 is not compatible with DNXCore,Version=v5.0 (win7-x64)" error installing xunit 2.1.0-beta4-build3109 into a new PCL project on Win8.1.

**Comment by bradwilson** [comment]
@jonorossi That's a bug in the UWP tooling where it doesn't realize that ASP.NET 5 is compatible with PCL 259 libraries (it has nothing to do with xUnit.net).

There are two ways to address the issue:

1. Uncheck "ASP.NET 5" in the PCL target selection list and everything compiles, and continues to work just fine on ASP.NET 5.

2. Alternatively, you can stop targeting the "meta" packages and instead just directly take assembly references as needed (since the bug is in the meta package). This is the route we chose to go down for xUnit.net (for other reasons, but I also think it's not a terrible idea for people to know exactly what they're pulling in and why, given that every reference means larger deployments). Once you've done this, the PCL targets checkbox agrees that you are successfully targeting ASP.NET 5. Example project.json file: https://github.com/xunit/xunit/blob/master/src/xunit.runner.reporters.DotNetCore/project.json

**Comment by jonorossi** [comment]
@bradwilson many thanks, I suspected it was something broken with all this new stuff but the logs were pretty useless in working out why. I'll give the second option a go tomorrow.

**Comment by MattGal** [comment]
@JonRossi, sorry for the slow response, but it has been my turn to be sick this week unfortunately.  I am back in the office this morning.
Some comments:
-	You’ve already figured it out, but yes enabling a PCL for .NET Core unfortunately requires installing the Windows SDK.   I can add this to a readme somewhere, it’s definitely no obvious.
-	There are two reasons I chose to use an ASP.NET 5 class library for the test project:
o	As you’ve already noted, currently you can’t add a reference to even Xunit to a PCL targeting .NET core 5.0.  This is due to a disparity in the list of frameworks that the project types think they are compatible with.  DNX, the host when building ASP.NET Class libraries, has some magical goo that lets it decide that “portable-net45+win+wpa81+wp80+monotouch+monoandroid+Xamarin.iOS” is compatible with .NET Core, while the PCL project system does not.  I am following up with the product team regarding this.
o	Even if you could add XUnit to a PCL that targeted Core, there is no mechanism currently for executing these tests in the context of a CoreClr AppDomain, i.e. they would be running the right code but under the desktop framework.   There are definitely slight behavioral differences between desktop and Core, and I wouldn’t feel comfortable trusting results of tests executed against a different framework than was used.
If you can send logs from the output window,  the results of “dnvm list” and “dnu restore” from a developer command prompt, I can probably help with your issues building the unit tests today.  You’ll want 1.0.0-beta5 to be both your default (achieved by “dnvm alias default 1.0.0-beta5”) and active runtime version when doing this (“dnvm use –p 1.0.0-beta5”).  If you use google hangouts, you can ping me directly via galbraith.matt@gmail.com for help with this, as the ASP.NET 5 project system tends to be a bit finicky.

**Comment by bradwilson** [comment]
@MattGal wrote:

> Even if you could add XUnit to a PCL that targeted Core, there is no mechanism currently for executing these tests in the context of a CoreClr AppDomain, i.e. they would be running the right code but under the desktop framework.

This is true for old-style PCLs, too.

We have platform-specific test runners for all three target frameworks in `dotnet` (`net46`, `uwp10`, and `dnxcore50`). You should be able run a `dotnet` PCL with a desktop runner for `net46` (like xunit.console) or the DNX runner, and they will run in the appropriate context. For UWP, you'll unfortunately need to make a UWP-specific unit test project (as this is a requirement of the Visual Studio runner); then you should be able to just link all the existing test files into it and run them on all the supported UWP platforms.

If you were doing this in PCL259-land, you'd have a very similar list of restrictions, and end up with several unit test projects (which shared the actual test code) in order to be able to run those tests on all the supported platforms.

**Comment by MattGal** [comment]
Thanks for the pointers @bradwilson  

While I recognize how things worked for old-style PCLs (which didn’t explicitly include Core support), I thought it’d be nice to set them up so they did execute in a hosted CoreCLR environment to be as close to realistic as possible.  There are definitely some slight behavioral differences in APIs that otherwise match up on the frameworks, and if a particular library managed to accidentally depend on this difference, it’d be cool to catch this in testing instead of production.

That said, I’m certainly open to more specific guidance about how to make XUnit tests for the new style PCLs.  I had played with this a little with no luck, but had not tried the Universal Windows Unit Test project type until today.  I note that the tests do run in a reference .NET Core environment, so that’s great, but when I naively take a new UnitTestProject and add XUnit to it, I do not actually see my methods I’ve decorated with [Fact] attributes get called when I run tests.  Is there more I need to do with this project type to get that working?

Since @jonorossi is refactoring tests and the tests I’ve provided are indeed out of date, I’d be happy to see what Jonathan serves up and help to mold it into a UWP unit test project instead, I’m just unsure how XUnit fits in.  (I am trying with 2.1.0-beta4-build3109 )

**Comment by bradwilson** [comment]
@MattGal The point is, there's no such thing as a single "hosted Core CLR environment". The only two Core CLR environments today are massively different: UWP and DNX.

If you're just talking about general environmental compatibility, then 98% of that is selected for you with the profile; the rest, you'll have to discover via exploratory testing.

**Comment by MattGal** [comment]
@bradwilson The hosts are massively different sure, but while the executables may be different (UnitTestProject.exe versus dnx.exe, for instance) the framework assembly implementations used come from the same source code and are built the same way, thus I would expect them to share any peculiarities versus desktop .NET.  

Either way, since you know a lot about XUnit, can you please comment on the best project type and any setup I need to do for unit testing this library?  Simply making it a desktop library is feasible, and I'd expect that to just work, but if you can share best practices for this very new project type that'd be great.

**Comment by bradwilson** [comment]
> while the executables may be different (UnitTestProject.exe versus dnx.exe, for instance) the framework assembly implementations used come from the same source code and are built the same way, thus I would expect them to share any peculiarities versus desktop .NET.

That's actually not true.

At the very least, Desktop .NET will be different from Core CLR (which will be different from Mono, which will be different from Windows 8.x, which will be different from Windows Phone 8.1, etc.). The contracts swap out implementations at runtime which can wildly different.

**Comment by MattGal** [comment]
You are right, although running this library on Desktop was never my intent as I assumed users of the package would get the desktop implementation if so.  This work started mostly because ASP.NET 5 really wanted Moq, and Moq needed Castle.  Thus, while I understand the radical differences between implementation of desktop and core versions I expected this library on desktop to be unsupported (or at least less desirable, as it has fewer features).

After conferring with some folks, I learned that to properly test these scenarios, we'd need to have tests that use the Universal App Unit test type (using the .NET Core that Universal apps do, which has identical implementation to the dnxcore50 versions in some places but not all) AND as a ASP.NET 5 Class library using the implementation assemblies it fetches from Nuget.  However, I think it's fairly safe to pick one and stick to it, until one of these differences breaks something.    I'm happy to do the leg work to pursue either of these test project types once @jonorossi weighs in.

**Comment by jonorossi** [comment]
@MattGal sorry for the delayed reply, your conversion with @bradwilson scared me off looking at this until I actually had some decent time to look at it. I'll take a bit of a look tonight, and hopefully have an answer tomorrow.

**Comment by MattGal** [comment]
I didn’t mean to scare anyone, I simply was trying to point out that from a realism standpoint running tests against the desktop framework versus one of the two (UWP or DNX flavors, which share a lot of their implementation assemblies but not all) CoreClr ways to do it has some differences and should be considered.

I don’t mind whichever way the Castle Project team chooses to go, I just would like to get this done so I can work on Moq4 on Core support in earnest, and am happy to help out any way I can towards that goal.

**Comment by jonorossi** [comment]
@MattGal I was only half joking, and it wasn't at all about which framework to run against, but the fact that this is all so complicated and messy for new tools. There was never a question about which framework the tests would run on, we're testing DynamicProxy which pushes runtimes to their limits (i.e. we've crashed the desktop CLR in the past and probably still can, and some of our unit tests are disabled on mono because we crash the mono runtime, we need to expose those bugs). I just thought there would be a fairly simple way to run xunit.runner.dnx against a PCL without going via the project system.

**Comment by MattGal** [comment]
As of the present time, this is the recommended approach for having tests execute under DNX unfortunately.  You might be able to refactor the actual test methods into another PCL and have a smaller DNX project wrapper, but it's still going to need a basic xproj setup as seen in the PR for the time being.

Getting feedback about issues like this is a big motivation for doing this work.  I greatly value your feedback about why this is challenging / irritating, as well as your ideas for how things could be better.

**Comment by jeremymeng** [comment]
As @MattGal switched to focus on other projects, I am picking up his work in this space.  Please keep providing feedback.

**Comment by richlander** [comment]
We'd love to get this PR merged so that we enable Castle.Core and it's upstream users (e.g. Moq) to support .NET Core. It would be great to understand what is needed to make that happen. We're happy to update the PR, as needed.

I read through the last several posts. The main concern seems to be around the environment to run the tests in. I would think that enabling a "generic .NET Core host" is not the right direction. It's interesting to get base-level feedback, but not the actual environment that the code will be running in. That seems to be what @bradwilson is arguing. How about we enable this for ASP.NET 5 first (to get something actually working and get feedback) and then add support for Win 10 UWP and raw .NET Core.

Thoughts??

/cc @jonorossi, @bradwilson, @kkozmic, @jeremymeng

**Comment by richlander** [comment]
BTW: Super happy to organize a conf-call/skype/facetime on this so that we can come up with a plan.

**Comment by jonorossi** [comment]
> I would think that enabling a "generic .NET Core host" is not the right direction.

Unless I'm missing something I've got no idea why you thought anyone was asking for a generic .NET Core host, I'm definitely not. The discussion came from the PR being changed from a project.json to a PCL.

> There was never a question about which framework the tests would run on, we're testing DynamicProxy which pushes runtimes to their limits (i.e. we've crashed the desktop CLR in the past and probably still can, and some of our unit tests are disabled on mono because we crash the mono runtime, we need to expose those bugs). I just thought there would be a fairly simple way to run xunit.runner.dnx against a PCL without going via the project system.

Now to where things are at. In August I introduced (b11f8ca4b52319d99fef3946a2096631642c959d, 04f2e853303cc7a86a7b2a4302cb425d5d1ddbba) the `FEATURE_XUNITNET` symbol, an NUnit shim over xUnit.net, and made heaps of changes to unit tests to rely less on the unit test library. The master branch has those changes to build and run unit tests on either library. I did this work with .NET 4.5, however we won't be supporting running the unit tests for 4.5 on xUnit.net.

I then started looking at the PCL in the pull request. I played with the new DNX tools that wrap an existing MSBuild project and found it was too hacky having the library as a PCL and the unit test project a project.json based project, so created both as project.json projects to test that out, which you can see on the `netcore` branch (427b489bcbb5903b468334e70696a6f209a1189d), I also played around to work out how the CI build would handle these (http://builds.castleproject.org/viewType.html?buildTypeId=CoreNetCore_MasterNetCore).

You can see in that branch that the project has very few files just to test things out, however during late August I then got through about half the codebase adding new conditional compilation symbols for functionality missing from .NET Core (these changes aren't pushed to any public repo). By the end of August I had to drop things to work on paid work and haven't got back.

The reason I'm not using the changes in the PR is because I've asked a few times for us not to introduce yet another platform symbol. I did the work to introduce `FEATURE_LEGACY_REFLECTION_API` and `FEATURE_SERIALIZATION` which massively reduced the original PR littered with `#if NETCORE`, and then got rid of all use of `MONO` from the library (unit tests still have a few `__MonoCS__`). I'll be dropping Silverlight support soon, and want to sort something out with the few `DOTNET35`, `DOTNET40` and `DOTNET45` symbols. I really didn't think this was a big ask since MSFT employees are doing the same thing in other projects like MSBuild, but it and the unit test problem was just ignored with continued comments about what is needed to merge this as-is. Doing this work also exposed a bunch of places `#if NETCORE` wasn't needed that the PR has it because it forces you to wrap the code with the "reason", the more conditional code the more pain we have to maintain.

> the changes definitely weren't small like they are now until I introduced the FEATURE_SERIALIZATION and FEATURE_LEGACY_REFLECTION_API conditional compilation symbols into master. I've hinted at it a couple of times, but my push is really to avoid another platform conditional compilation symbol and so I'm trying not to just rush a merge that'll just give us more technical debt.

Right now I'm still working on paid work and about to go on holidays, but the breaking changes and general flakiness I've read others talking about over the last month hasn't made me excited to get back to it.

**Comment by richlander** [comment]
Thanks for the indepth explanation, @jonorossi . That's helpful. I get why you don't want to add another platform symbol. Your approach makes sense, particularly if it results in less code being conditional.
 It sounds like you've done a lot of work. Thanks for doing that. 

Is it possible for you to push that branch? We'd love to help with that. We are not attached to the work we've done and can help move your branch forward towards the goal of supporting .NET Core. We do have a dependency on Moq and really want to see that library available on .NET Core.

What would you like us to do next? We'd be happy to create a PR on top of your work, If you are not going to be available to move your work forward for a while.

@MattGal said that he was creating a PCL since it provided a better developer experience in VS. What are your thoughts on that?

**Comment by richlander** [comment]
FYI: We're going to start looking at the NUnit port, too.

**Comment by jonorossi** [comment]
Obviously I wasn't able to get my changes pushed to a branch before I went on holidays, my VS2015 expired (I guess it was in trial mode) and it failed to activate both times I tried. It really baffles me why MSFT decided to require I sign into VS2015 with my MSDN account to use it (it doesn't support license keys anymore), it blocked me from using it because it wouldn't activate and I know others that develop on a network without internet access.

I'm back in less than 2 weeks and I'm likely going to have a bit of time during that time and will see what I can get sorted.

**Comment by richlander** [comment]
@jonorossi  Great. Looking forward to seeing your changes. A lot of other projects would benefit from that. We can help once the code is pushed.

NUnit is coming along nicely: https://github.com/nunit/nunit/pull/900. Not sure if that is still relevant to you.

Ya, the license thing is a pain. I was talking to @jeremymeng. He said that there is a text box to insert a product key for VS. Do you have one? Have you tried that?

**Comment by bradwilson** [comment]
FYI, looking at what NUnit is doing, their `netcore` support appears to be .NET 4.6 (which means no Mono support).

**Comment by jonorossi** [comment]
@richlander yep, I see the text box to enter a license key, however MSDN won't give me a license key for Professional or Enterprise edition like I've got for VS2013 Ultimate.

> A product key is not offered with this edition of Visual Studio. To unlock the product, you must sign in using the Microsoft account associated with your active MSDN subscription. By signing in, your IDE settings will sync across devices, and you can connect to online developer services.

Others have the same pain: http://stackoverflow.com/questions/31599844/how-is-visual-studio-2015-licensed-for-msdn-subscribers

**Comment by bradwilson** [comment]
I have two MSDN subscriptions (work and OSS), and they both let me use
product keys (I've used them with Pro and Enterprise VS 2015).

Is your MSDN subscription fully paid, or does it come through some other
source (like BizSpark)? My guess is that, if it's the latter, that's why
they won't give you a "forever" license.

**Comment by jonorossi** [comment]
@bradwilson yep, BizSpark. I've always used my employer's VS license for open source development over the last decade (BizSpark now, but Premium or Pro in the past). Just looked up the 16 Jul 2015 BizSpark email informing that VS2015 will be Microsoft Account only, and that VS2013 will change on the 20th to Microsoft Account only. I just assumed that change was for all Visual Studio users since nothing else under BizSpark changed like that and this part of the email (emphasis mine):

> BizSpark members who complete the graduation process after 3 years in the program can keep most of the software downloaded through the program. However, **because all versions of Visual Studio 2013 and 2015 will require you to sign in using a Microsoft account** associated with an active MSDN subscription, startups will need to renew their MSDN subscriptions for each user running Visual Studio after graduation.

Signing in this time worked which will have to do, even thought I don't like that VS has access to production Azure.

**Comment by jeremymeng** [comment]
@jonorossi Hope you had a wonderful vacation!  Any updates on pushing your changes so we can help working on top of it?

**Comment by Eilon** [comment]
Hi folks, @Eilon here from the ASP.NET team at Microsoft. Recently @pranavkm on my team started using a forked build that @jeremymeng was working on in order to enable using Moq in DNX/ASP.NET 5 applications. In order to use Moq we need to upload the forked build to our MyGet feed.

I wanted to let people here know we're going to use the fork in the meantime just in case anyone felt that would have any issues. Once the official packages are out we'd of course switch to that.

Any objections to use using the forked build on our MyGet feed?

**Comment by jonorossi** [comment]
@jeremymeng sorry for not replying, I must have missed your comment in the thousands of GitHub notifications.

@Eilon go for it, can't imagine any reason this would cause problems, it's your private MyGet feed at the end of the day.

Last week I pushed my newer changes into the netcore branch which I started in August. I also spent heaps of time last week porting heaps of code so that .NET 4.5 and .NET Core code are as much the same as possible, look at the description of each conditional compilation symbol in README.md. My goal is to remove as many of those symbols as possible by the end of the port. I've excluded DictionaryAdapter completely for now and have still got heaps of work left to do on the unit tests (however I implemented the NUnit shim over xUnit back in August) and had them all running on .NET 4.5.

These are the issues I've logged so far under the corefx project for missing APIs, there are many more to come, however I get the feeling the team isn't going to add much more for 1.0.
* https://github.com/dotnet/corefx/issues/4462 - Expose GetMethod and GetConstructor extensions with both BindingFlags and Type[]
* https://github.com/dotnet/corefx/issues/4463 - Expose System.Reflection.TargetException
* https://github.com/dotnet/corefx/issues/4491 - System.Reflection.Emit.AssemblyBuilder.Save

The branch and TeamCity builds:
* https://github.com/castleproject/Core/tree/netcore
* Diff: https://github.com/castleproject/Core/compare/master...netcore
* http://builds.castleproject.org/project.html?projectId=CoreNetCore

Feel free to send a PR for incremental changes to that branch. I'll be continuing to work on it over the next few weeks, however I just can't put it before client work anymore.

**Comment by richlander** [comment]
Thanks @jonorossi! That's great. I sent mail to folks on my team to take a look. I'm hoping they can help you land this PR.

**Comment by jeremymeng** [comment]
@jonorossi Thanks for pushing your changes!  I will start contributing to your branch.

I noticed that in the `netcore` branch, you put the `project.json` file in the same folder as the other projects.  It would caused problems.  I believe with UWP tool installed, VS 2015 will try to use `project.json` to restore packages for **ANY** other projects in the same folder. For example I am getting an error for Castle.Core project
```
1>------ Build started: Project: Castle.Core, Configuration: NET35-Debug Any CPU ------
1>Build started 11/17/2015 10:52:48 AM.
1>C:\Program Files (x86)\MSBuild\Microsoft\NuGet\Microsoft.NuGet.targets(109,5): error : Your project is not referencing the ".NETFramework,Version=v3.5" framework. Add a reference to ".NETFramework,Version=v3.5" in the "frameworks" section of your project.json, and then re-run NuGet restore.
1>
```
As far as I know there are two options to avoid the issue

1. Put the dnx project in its own folder.  Since the output assembly has the same name as the folder where the `project.json` is, we would have to put it somewhere like `src\Castle.Core.NetCore\Castle.Core` then explicitly include the source files in `project.json`.  The VS experience of this approach however will be worse, as the files will not be listed in the solution explorer, search doesn't work, etc..
1. Use the modern PCL project.  Matt's PR took this approach.  
    - The project file can be put in the same folder as other projects.
    - All the VS features remain the same.
    - It might seem complicate having to wrap the output dll before using it in the dnx test project, but it is a supported way, and in VS it is just the same as adding other kinds of project-to-project references.

I have done both, and if possible, I personally prefer the second option for its benefits.

Please let me know what you think?  I am happy to make the fix using either one, or any third approach.

**Comment by jeremymeng** [comment]
@jonorossi oh, and the NUnit team has released 3.0 with core support (still lacking a full feature console runner and Visual Studio integration.)  Please let me know if you want to use that.

**Comment by jonorossi** [comment]
@jeremymeng I was using VS2013 to open the .NET 3.5/4.0/4.5 csproj files, so I assume those UWP tools only cause problems in VS2015. Using a PCL for the libraries and then DNX for the unit tests project is inconsistent, why do you prefer using two different project formats? Matt, Brad and I already spoke about doing this earlier in this issue. Why does it work for James in [JSON.NET](https://github.com/JamesNK/Newtonsoft.Json/tree/master/Src/Newtonsoft.Json)?

I had been watching the NUnit issue but I don't know what the NUnit 3.0 support can do in .NET Core. If it doesn't have a unit test runner that actually runs on .NET Core then it is pointless for us, as previously discussed in this issue the unit tests must run on the actual runtime it is supported on. Do you have more details on NUnit?

**Comment by jeremymeng** [comment]
> Why does it work for James in JSON.NET?

I think he fixes it by adding `net35`, `net40`, etc. in the `project.json`.  We could do similar too.


> Do you have more details on NUnit?

Right now they support running NUnit tests in a dnx console project.  Please see a blog post from one of the maintainers (http://www.alteridem.net/2015/11/04/testing-net-core-using-nunit-3/)  for more information.

**Comment by bradwilson** [comment]
> Using a PCL for the libraries and then DNX for the unit tests project is inconsistent, why do you prefer using two different project formats?

DNX unit tests are an application, not a class library. The entry point for that application is provided by the unit test framework (which requires DNX, not PCL, because it requires access to DNX-specific features).

It wasn't my idea to make the test runner a reference (ask @davidfowl), which is why you're in this whole nonsense problem to begin with.

**Comment by jhogins** [comment]
@jonorossi Random lurker here. @bradwilson is correct, NUnit on .NET Core requires an application, as there is no runner yet and I believe dynamic assembly loading is not allowed anymore. I have had success using a barebones UWP application and NUNitLite (packaged with NUnit 3.0) to run the tests from VS. We haven't been able to figure out how to run it from the command line.

**Comment by jonorossi** [comment]
@jeremymeng JSON.NET has those renamed project.json files, but there is also a file named just project.json.

@bradwilson thanks, I'm fully aware the unit tests are an application, but I still maintain the position that the shipping libraries might as well be DNX/project.json libraries rather than a PCL if we aren't actually shipping a PCL because there isn't a profile that we can actually use, and if we are maintaining csproj files for the .NET Framework.

@jhogins thanks, do you know what the limitations of NUnitLite are compared to the full console runner? It looks like 3.0.0-rc added support for the test results XML file which we need for TeamCity.

**Comment by jhogins** [comment]
We have not seen any API limitations when running NUnitLite. The only differences we have found are in the supported command-line arguments. Here is the help text from NUnitLite in NUNit 3.0.1, which is the only correct documentation we've found:

Usage: NUNITLITE [options]

Runs a set of NUnitLite tests from the console.

Assembly:
      An alternate assembly from which to execute tests. Normally, the tests
      contained in the executable test assembly itself are run. An alternate
      assembly is specified using the assembly name, without any path or.
      extension. It must be in the same in the same directory as the executable
      or on the probing path.

Options:
      --test=NAMES           Comma-separated list of NAMES of tests to run or
                               explore. This option may be repeated.
      --where=EXPRESSION     Test selection EXPRESSION indicating what tests
                               will be run. See description below.
      --timeout=MILLISECONDS Set timeout for each test case in MILLISECONDS.
      --seed=SEED            Set the random SEED used to generate test cases.
      --stoponerror          Stop run immediately upon any test failure or
                               error.
      --wait                 Wait for input before closing console window.
      --labels=VALUE         Specify whether to write test case names to the
                               output. Values: Off, On, All
      --verbose, -v          Display additional information as the test runs.
      --help, -h             Display this message and exit.
      --version, -V          Display the header and exit.
Notes:
    * File names may be listed by themselves, with a relative path or 
      using an absolute path. Any relative path is based on the current 
      directory or on the Documents folder if running on a under the 
      compact framework.

    * On Windows, options may be prefixed by a '/' character if desired

    * Options that take values may use an equal sign or a colon
      to separate the option from its value.

**Comment by bob-devereux-zocdoc** [comment]
Is there a timeline on how long coreCLR compatibility will take?

**Comment by jonorossi** [comment]
@bob-devereux-zocdoc no, there is no timeline. @jeremymeng and I are working though #119 at the moment, but there is still more work after this PR.

**Comment by rjdevereux** [comment]
Thanks! Looking forward to trying to migrate some projects to core.

**Comment by adamralph** [comment]
I see that https://github.com/castleproject/Core/pull/118 is marked against the 4.0 milestone. Is .net core support going to be part of 4.0?

**Comment by jonorossi** [comment]
@adamralph yep, part of v4.0 as there is definitely enough code thrash to warrant a major release.

**Comment by adamralph** [comment]
@jonorossi do you follow SemVer? Is it 4.0 because of breaking changes?

**Comment by jonorossi** [comment]
@adamralph technically there are breaking changes to some public classes that are in the `Castle.Core.Internal` namespace that people shouldn't have been using, however I think Windsor does. I want to bump the log4net version (#109) which will drop support for 1.2.10 (so also a breaking change) because of the breaking log4net API (not just the change in strong name key), and other changes (e.g. #77) also may have unexpected behaviour change for some users. Other than that just generally there has been so much code change from .NET Core work that the risk of regressions is higher.

In the past I haven't been following SemVer to the letter, I've released some behaviour changes as patch version bumps that are bug fixes that are technically breaking changes. I understand the SemVer concept in spirit, but IMHO it is fundamentally flawed because just about every code change can be a breaking change for some user.

**Comment by adamralph** [comment]
@jonorossi ok, makes sense. Thanks for the info.

**Comment by adamralph** [comment]
It seems like the bulk of this work has been completed - https://github.com/castleproject/Core/pull/125#issuecomment-183204026

Speaking as an owner of a downstream package (FakeItEasy), one of the most useful things that could happen right now, is if a pre-release version of Castle.Core were pushed to NuGet, containing the .net core assemblies. This would allow us to get to work on supporting .NET core in FakeItEasy, and push our own pre-release packages containing .NET core assemblies.

**Comment by xied75** [comment]
@richlander Got a quick question as UWP is mentioned multiple times in the thread, I wonder how is that possible for DynamicProxy to be supported in UWP? Given the current .NET API allowed there, it doesn't look like you can emit new types? So when you say UWP support to follow, do you mean the .NET API surface could be extended?

Related but off topic to this issue is that who actually control what is available to UWP sandbox? Is that Windows team or .NET team?

**Comment by shanselman** [comment]
I agree with @adamralph a pre-release version would unblock a number of people @jonorossi

**Comment by hammett** [comment]
If this is a blocking issue, feel free (and blessed) to fork Castle.Core. @kkozmic, @jonorossi and myself seem more occupied by other activities. I have no interest in dealing with this codebase in my spare time, at least not for free. Please slightly change the name of the final assembly to indicate it's a fork and possibly not compatible.

Another route would be to make sure the PR(s) wont break Windsor, and they will be promptly accepted. 

Thanks

**Comment by thecodejunkie** [comment]
@hammett @kkozmic @jonorossi is this the official stance from the Castle.Core team, i.e

>  "We won't have time / the desire to make Castle.Core compatible with the CoreCLR in the foreseeable future so we would rather see a community fork"

The reason I ask is because your comment does not come across as an official stance from the team and I think it is important that it is clear because, as @shanselman and @adamralph points out, there is a downstream community blocked.

Would the be any kind of timeframe when this would be done by the team? Forking is not going to be as clear cut (strong names and what not) and it would add an even larger burden on the community since that fork would have to keep pushing out mirror releases to be feature compatible and make sure bugs are removed if they have been reported and fixed in upstream.

**Comment by hammett** [comment]
It's unlikely that there will be much activity on Castle.Core that will cause added burden to keep the forks in sync. I dont want to speak for @jonorossi but seems that he -- as myself [who some might not know, originally wrote DynamicProxy way back when] -- doesnt have much time for this, So let's take the noble action of stepping out of the way.

I'd be willing to even give repo write access to whomever is sending the PRs if that would streamline the process and avoid the fork, as long as source code/behavior compatibility is preserved for Windsor.

I just dont want this to be a bottleneck as it has been for the past several months. I have no immediate interest in .net core or much of .net to be frank but there's no point in blocking this further.

**Comment by adamralph** [comment]
> So let's take the noble action of stepping out of the way.

:clap: that seems like the right thing to do. The other side of that is that some new people need to step up and take the baton as join as new maintainers.

> I have no immediate interest in .net core or much of .net to be frank

Nothing personal, but that seems like strong grounds to step down as a maintainer of the project (which, perhaps is what you are alluding to).

**Comment by hammett** [comment]
I havent been a "maintainer" since 2012 - my lack of participation on this and other threads should have made this fact abundantly clear.

@jonorossi has been doing a fantastic job, but OSS is always second (or 3rd, or 4th) to whatever pays the bills. And you dont see a lot of people willing to step in. A PR is not a takeover.

**Comment by rjdevereux** [comment]
First off, all the interest in this thread shows how important this project is to the .Net community, directly and indirectly a lot of people are dependent on it.  Thanks for all the work you guys have done building and maintaining it.

There have been offers to help from the MSFT guys, if you look back through the posts you can find @richlander offering to help organizing communication and pledging support from MSFT developers, I'm hoping those offers still stand.  But they were also cautious and wanted to respect that they don't own the project.

Sorry if I'm putting words in other people's mouths, but I think there is a lot of excitement to keep this moving forward and what folks are looking for is a little direction on how to move forward from the maintainers, I think there are people ready to help get this working on Core, they just need direction or permission.

**Comment by bob-devereux-zocdoc** [comment]
Am I right at pull [125 ](https://github.com/castleproject/Core/pull/125 )is where this is blocked right now?

**Comment by hammett** [comment]
I'm aware. 

Directions were given, but repeating: the PR as stands is not fully compatible, so either fork Castle.Core with our blessing or provide a PR that doesnt break Windsor. Nothing more, nothing less.

**Comment by jonorossi** [comment]
Guys please relax, this is getting heated and a little disrespectful for no good reason.

I'll start by saying that @hammett doesn't speak for me (as he did mention), and I do not share the same opinions as he does (which he is definitely entitled to).

I've been involved with the Castle Project for the last decade and a maintainer for majority of that time and until recently I've never encountered so many demands (not just .NET Core related) and so little willingness to actually help. I've never been paid for a single hour of time or ever worked on an employer's time, it has been all out of love. If it wasn't obvious to those in the community the Castle Project has been suffering badly from having so few maintainers over the last few years, and ironically Microsoft has dealt the biggest blows to our community since our highs in 2007-2010 with each new project they released killing one of our own and shrinking our community.

Personally I've been mostly ignoring Castle stuff for the last few weeks because I have had a lot going on in my life, I've recently lost a non-immediate family member, also lost one of my best friends to suicide, been massively behind on consulting work and swamped, and just this week moving out of the office I sublease because it is closing with short notice.

I don't want to drone on but I've put in a couple hundred hours already to get Castle Core ported and supportable on .NET Core, and stated here in this issue (https://github.com/castleproject/Core/issues/90#issuecomment-165320289) I was waiting for RC2 to be released which no one even commented to say had slipped from release early February to now having no date at all. If you don't think I was committed to get this out and don't think the work I was doing was worthwhile then I might as well give up and walk away from all of this, my efforts are obviously not appreciated.

+1s and more demands don't help things move along. #125 wasn't actually finished (but @hammett has now merged), I had asked for the PR to include all the other AssemblyInfo attributes which we'd still need to make a release (https://github.com/castleproject/Core/pull/119#issuecomment-168586441), I don't know why everyone was jumping on one tiny PR.

Replies to comments since my last comment from Jan 5:
* @thecodejunkie I don't like that you made your comment look like a quote from @hammett, but no @hammett's stance isn't the Castle Project's official stance, what's left of our PMC haven't ever discussed the topic, if it wasn't clear from @hammett's comment I'm really the only one keeping the lights on lately.
* @adamralph no need for :clap: and disrespect, @hammett hasn't been standing in anyone's way to contributing.
* @rjdevereux there has been help from Microsoft, I've been working with @jeremymeng over the past 6 months, no disrespect to him but I've had to hold his hand on each PR to ensure the changes keep the codebase maintainable. As I said above in this comment the PR wasn't finished, I just didn't explicitly state that because I was sidetracked.
* @bob-devereux-zocdoc obviously not now since @hammett just merged it, I still don't know why everyone got hung up on #125, maybe because I went dark, but there is still plenty more and I've been saying that for a while in all the different threads but I feel like the only one looking at the future of this project not the quickest way to hack everything to get a pre-release build out.

Where to from here? My opinion of DNX RC1 was that it was flaky and Microsoft was barely supporting it, but now that RC2 has been delayed until sometime in the future what do people want to do? Remember the Castle Project isn't a vendor, you don't pay anyone, you guys are the community, you have to help decide where we go, you have to get involved, I don't want to hear anymore demands, I want solutions. This issue has become long and its final purpose will be to create new issues of remaining things people can work on.

**Comment by hammett** [comment]
hey @jonorossi . I jumped in exactly coz you seem to have gone dark. Of course your involvement is appreciated. I merged that small PR exactly coz it's small. I just want this "blocking" status out of the way. Since I dont see any future release happening, IMHO the risk is very low.

**Comment by jonorossi** [comment]
@hammett no problem, my fault. With every new Castle Project notification I dreaded coming back to any of this.

**Comment by hammett** [comment]
This is not an easy codebase to deal with. That may partially explain the lack of people stepping in and changing code. 

I even remember as a MSFT, suggesting to the .net team to implement the same DP support -- as it's such valuable functionality that so many projects depend on -- with the welcome side effect of unburdening us, that back in 2009/10. I wish they had listened. 

And from what it seems, they regret that too.

**Comment by jonorossi** [comment]
@hammett I'd love all the runtime bugs to be Microsoft's problem, that way they'd probably get fixed, I know @kkozmic loved getting knocked back on MSConnect every time :). Remember the pain of .NET 3.5 SP1.

**Comment by thecodejunkie** [comment]
@jonorossi first of all let me pass on my condolences for the recent events that have taken place in your live, I truly mean that. I faced a chain of events last year which made me minimize my own commitments on thins that were not essential for my daily life.

I think there's been a chain of misunderstandings across the board here. Nobody is trying to drum up and drama not point any fingers. I never, intentionally, made anything look like a quote by @hammett I paraphrased and in fact, repeatedly, asked if I had interpreted his remarks correct, that is all. Nothing more, nothing less. No ill intentions.

> +1s and more demands don't help things move along

My guess is that the +1 are not there to stress anyone (at least that's how I interpreted them), but rather highlight that this is something that is important for other people as well. Unfortunately GitHub has no upvote system in place, unlike GitLabs. But I digress. Castle.Core is a "close to the metal" component for many in our community (closed or open) and as suck when people start migrating their projects over to CoreCLR, there will be a ripple effect. I've seen it in the project I help maintain as well, which indirectly caused a ripple effect on Castle.Core (because of our usage of FakeItEasy).

Many of us here are open-source maintainers ourselves. I created Nancy 5+ years ago and have been part of the team that maintains it ever since. @adamralph have been maintaining FakeItEasy for many years now. Trust me, we _really_ understand what it means when you say that is is a labor of love. I've invested thousands of hours on my project alone and our main "competitors" are the entire ASP.NET team and their range of frameworks and tooling teams :) 

>  I don't want to hear anymore demands, I want solutions

The problem is that Castle.Core is a complicated project and you simply do not jump into a project like that and start pulling at the inner mechanics, there are simply too many moving parts to understand. What might be needed is if you guys could work _with_ the community and what ever contributors you have been having in the past. This means identifying problems and create issues that are up-for-grabs to the community. That is what we did with Nancy. https://github.com/NancyFx/Nancy/issues/2220 generally project maintains frowns upon someone sending massive "I changed a lot of stuff" pull-requests, so creating a list of known issues might aid contributions.

As good start could be to run the [.NET Portability Analyzer](https://visualstudiogallery.msdn.microsoft.com/1177943e-cfb7-4822-a8a6-e56c7905292b) on the project and extract issues from there. Our experience has also been that if you reach out to Microsoft they are quite willing to help with CoreCLR related questions and I've seen @jeremymeng do some work. 

We, the team at Nancy, are more than happy to help answer questions and share our experience with our journey on being compatible with CoreCLR. I doubt anyone would feel comfortable with sending deep pull-request and we are also currently working on our own CoreCLR story

> My opinion of DNX RC1 was that it was flaky and Microsoft was barely supporting it 

We are currently targeting rc1 and have had very little issues with the DNX runtime, the APIs or the tooling. And this is after having worked with it for over 40 hours per week for the past 3 weeks. The API changed between rc1 and rc2 aren't that big (I think @jeremymeng can provide details here) and it will be a lot of runtime / tooling changes. 

Hope this clears things up

Thanks

/Andreas

**Comment by adamralph** [comment]
@hammett 

> I havent been a "maintainer" since 2012 - my lack of participation on this and other threads should have made this fact abundantly clear.

Perhaps "maintainer" isn't the correct term, but you have commit rights and are exercising them (proven by your merge today of https://github.com/castleproject/Core/pull/125). This sends the message to the community that you are a person who is actively involved in the project but 

> I have no immediate interest in .net core or much of .net to be frank

suggests otherwise.

Again, there is nothing personal in this, my observation that the latter comment is strong grounds for stepping down as a maintainer ("committer" if you like), is purely about giving the community a sense of who is actively involved, has an interest in moving the project forward, can be polled for opinion, etc.

**Comment by adamralph** [comment]
> Guys please relax, this is getting heated and a little disrespectful for no good reason.

@jonorossi really? Personally I've done my best to stay polite and approach this constructively. Can you point me to any offending comments?

**Comment by adamralph** [comment]
> @adamralph no need for :clap: and disrespect, @hammett hasn't been standing in anyone's way to contributing.

@jonorossi again, this was not meant in a disrespectful way. Apologies if it was misconstrued, but I was genuinely applauding @hammett's stance on this. I have the greatest respect for it.

**Comment by adamralph** [comment]
:+1: @thecodejunkie https://github.com/castleproject/Core/issues/90#issuecomment-188866211 largely echoes my thoughts

**Comment by jonorossi** [comment]
@thecodejunkie thanks, that means a lot, 2016 has been a rollercoaster. Sorry I saw our names and the formatted quote syntax and then another set of quotes, it seemed intentional.

I really do appreciate your work, I've been using Nancy for a few years now (ever since MonoRail lost all its contributors) and you and others are running that project very well. Every time I've interacted with you guys or follow threads I've been envious how quickly you respond and get issues sorted. Since @kkozmic moved on to non-.NET work I've felt like I can't even keep up with new issues being created let alone actually fix issues myself, I'm not blaming anyone definitely not @kkozmic who has done heaps more than fair share on Castle.

Don't worry we went through that whole massive PR of hacky code that got the thing sort of compiling the middle of last year, which is why I spent hundreds of hours individually porting things across cleanly so we could continue maintaining the codebase on multiple runtimes with a whole pile of [conditional compilation symbols](https://github.com/castleproject/Core/blob/netcore/README.md#conditional-compilation-symbols) that make sense (well to me at least).

The parts I wanted help with are all the bits around the sides (including the build), and getting log4net, NLog and Serilog integrations upgraded. It looks like we're actually still using beta8, not even rc1, I recall I couldn't get the VS unit test runner to work with rc1 but I assume that is fixed now. Someone mentioned earlier in this thread that NUnitLite got shipped quite a while after I wrote an NUnit shim over xUnit.net, but I didn't hear back from anyone to contribute anything other than mentioning it. Just logged a bunch of issues for these (tagged up-for-grabs), it is now 3:23am so I'll have to find another chance to see what else is left to do.

In terms of getting things fixed, most of the issues I logged against coreclr/corefx are assigned to rtm, but the main one I wanted fixed (and I made that clear) got set to Future without any explanation of why it couldn't be done now even though the code from the desktop CLR is still in there just compiled out.

**Comment by adamralph** [comment]
@jonorossi to echo @thecodejunkie, your efforts are hugely appreciated. As an owner of several OSS projects I know how difficult it can be to give the time you want to to these things and how often it can be a thankless task.

One good thing here is that this dialog has started and the community is being made aware of the challenges being faced. Clearly more muscle is needed to make .NET core support and other things happen so I hope that we as a community can step up and put in the effort required. @hammett has indicated a willingness to bring more committers on-board (granting write access) and for me, that is a good path forward for the project. I realise how difficult it can be to find willing and able people for this, but again, I'm hoping the community will step up here.

**Comment by jonorossi** [comment]
@adamralph yep, I think I misread your emoji, one of the reasons I dislike them. It's late here (3:31am) after a crappy day.

We've always been willing to bring on new committers, we've just had no one stay around for long enough to get there for a long time.

For those that want to know about .NET Core stuff not our people problems, I've just created a bunch of issues to track further work, please participate in those issues, I am now closing this issue.

**Comment by shanselman** [comment]
Thanks folks, to all including @jonorossi @adamralph @hammett etc. Sorry this was a painful thread. Yes, things will get better after RC2. Do let me or @jeremymeng how we can help advocate and help. I'm working on other ways as well. @martinwoodward

**Comment by jonorossi** [comment]
I know there are a lot of people subscribed to this issue so I'll post it here so as many people notice as possible. I've just released `4.0.0-alpha001` from our netcore branch:
* http://www.castleproject.org/blog/2016/04/07/core-4-0-0-alpha001-release/
* https://www.nuget.org/packages/Castle.Core/4.0.0-alpha001

Please log any problems you find with this version as separate issues, thanks.

**Comment by Serjster** [comment]
Any updates on this? Specifically for .NET Core RTM?
